SQL Expressions: Docs: List all allowed functions (#104097)
grafana/grafana: docs/sources/panels-visualizations/query-transform-data/expression-queries/index.md

  - **backfill** with next known value
  - **fillna** to fill empty sample windows with NaNs

#### SQL

> The SQL expression engine is a new expression engine that allows you to transform the data returned from your data source queries using a MySQL-like syntax.

Reviewer: ```suggestion
The SQL expression engine is a new feature that allows you to transform data from your data source queries using a MySQL-like syntax.
```

##### SQL Functions

A subset of MySQL functions are supported in expressions, for two reasons:

> 1. We use [go-mysql-server](https://github.com/dolthub/go-mysql-server) to execute the SQL statements. It supports a subset of MySQL functions, as per the documentation [here](https://docs.dolthub.com/sql-reference/sql-support/expressions-functions-operators#functions-and-operators).

Reviewer: Avoid using "we" in documentation: https://grafana.com/docs/writers-toolkit/write/style-guide/style-conventions/#address-users-clearly.

Use "Grafana recommends" or similar instead.

> 1. We use [go-mysql-server](https://github.com/dolthub/go-mysql-server) to execute the SQL statements. It supports a subset of MySQL functions, as per the documentation [here](https://docs.dolthub.com/sql-reference/sql-support/expressions-functions-operators#functions-and-operators).

Reviewer: ```suggestion
1. SQL expressions are executed using [go-mysql-server](https://github.com/dolthub/go-mysql-server), which supports a subset of MySQL functions. For details, refer to [Functions and operators](https://docs.dolthub.com/sql-reference/sql-support/expressions-functions-operators#functions-and-operators).
```

> 2. We restrict the use of potentially dangerous functions.

Author: Should we include a call-to-action? Something like:
"Open an issue on the Grafana repo if you'd like a function added to this list"

In the Dolt docs they say this ([here](https://docs.dolthub.com/sql-reference/sql-support#:~:text=Any%20deviation%20from%20the%20MySQL%20manual%20should%20be%20documented%20on%20this%20page%2C%20or%20else%20indicates%20a%20bug.%20Please%20file%20issues%20with%20any%20incompatibilities%20you%20discover.)):
> Any deviation from the MySQL manual should be documented on this page, or else indicates a bug. Please [file issues](https://github.com/dolthub/dolt/issues) with any incompatibilities you discover.

> 2. We restrict the use of potentially dangerous functions.

Reviewer: ```suggestion
2. Grafana restricts the use of certain functions to prevent the use of potentially dangerous or destructive operations.
```

- `LOCATE()`
- `LOWER()`/`LCASE()`
- `LTRIM()`
- `MID()`/`SUBSTRING()`
- `ORD()`
- `POSITION()`
- `REGEXP_SUBSTR()`
- `REPEAT()`
- `REPLACE()`
- `REVERSE()`
- `RIGHT()`
- `RTRIM()`
- `SUBSTRING_INDEX()`
- `TRIM()`
- `UPPER()`/`UCASE()`

###### Date/Time Functions

- `DATE_ADD()`
- `DATE_FORMAT()`
- `DATE_SUB()`
- `DATEDIFF()`
- `DAY()`
- `DAYNAME()`
- `DAYOFMONTH()`
- `DAYOFWEEK()`
- `DAYOFYEAR()`
- `EXTRACT()`
- `FROM_UNIXTIME()`
- `HOUR()`
- `MINUTE()`
- `MONTH()`
- `MONTHNAME()`
- `QUARTER()`
- `SECOND()`
- `STR_TO_DATE()`
- `TIME_TO_SEC()`
- `TIMESTAMPDIFF()`
- `UNIX_TIMESTAMP()`
- `WEEK()`
- `WEEKDAY()`
- `YEAR()`

###### Type Conversion

- `CAST()`
- `CONVERT()`

###### JSON Functions

- `JSON_ARRAY()`
- `JSON_CONTAINS()`
- `JSON_EXTRACT()`
- `JSON_INSERT()`
- `JSON_KEYS()`
- `JSON_LENGTH()`
- `JSON_MERGE_PATCH()`
- `JSON_OBJECT()`
- `JSON_QUOTE()`
- `JSON_REMOVE()`
- `JSON_REPLACE()`
- `JSON_SEARCH()`
- `JSON_SET()`
- `JSON_TABLE()`
- `JSON_TYPE()`
- `JSON_UNQUOTE()`
- `JSON_VALID()`

###### Window Functions

- `DENSE_RANK()`
- `FIRST_VALUE()`
- `LAG()`
- `LAST_VALUE()`
- `LEAD()`
- `RANK()`
- `ROW_NUMBER()`

###### Operators

- `BETWEEN`

## Write an expression

> If your data source supports them, then Grafana displays the **Expression** button and shows any existing expressions in the query editor list.

Reviewer: ```suggestion
If supported by the data source, Grafana displays the **Expression** button and displays any existing expressions in the query editor list.
```
